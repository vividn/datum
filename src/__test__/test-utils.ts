import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
