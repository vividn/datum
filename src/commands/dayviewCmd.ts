import { ArgumentParser } from "argparse";
import { dbArgs } from "../input/dbArgs";
import { MainDatumArgs } from "../input/mainArgs";
import { dayview } from "../dayview/dayview";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { connectDb } from "../auth/connectDb";

export const dayviewArgs = new ArgumentParser({
  add_help: false,
  parents: [],
});

dayviewArgs.add_argument("--n-days", "-n", {
  help: "number of days to show",
  type: "int",
  dest: "nDays",
});

dayviewArgs.add_argument("--start-date", {
  help: "start date to show",
  type: "str",
  dest: "startDate",
});

dayviewArgs.add_argument("--end-date", {
  help: "end date to show. Defaults to today",
  type: "str",
  dest: "endDate",
});

dayviewArgs.add_argument("--width", {
  help: "width of the svg",
  type: "int",
});

dayviewArgs.add_argument("--height", {
  help: "height of the svg",
  type: "int",
});

dayviewArgs.add_argument("--day-height", {
  help: "height of each day. If not specified, it will be calculated",
  type: "int",
  dest: "dayHeight",
});

dayviewArgs.add_argument("--output-file", "-o", {
  help: "output file, should have a .html or .svg extension",
  type: "str",
  dest: "outputFile",
});

dayviewArgs.add_argument("--watch", "-w", {
  help: "watch for changes",
  action: "store_true",
  dest: "watch",
});

export const dayviewCmdArgs = new ArgumentParser({
  description: "View the day",
  prog: "datum dayview",
  usage: `%(prog)s`,
  parents: [dayviewArgs, dbArgs],
});

export type DayviewCmdArgs = MainDatumArgs & {
  nDays?: number;
  startDate?: string;
  endDate?: string;
  width?: number;
  height?: number;
  dayHeight?: number;
  outputFile?: string;
  watch?: boolean;
};

export async function dayviewCmd(
  args: DayviewCmdArgs | string | string[],
  preparsed?: Partial<DayviewCmdArgs>,
) {
  args = parseIfNeeded(dayviewCmdArgs, args, preparsed);
  const db = connectDb(args);

  if (args.watch) {
    console.log("watching");
    const emitter = db.changes({ since: "now", live: true });
    while (true) {
      console.log("redrawing");
      await Promise.all([
        dayview(args),
        // redraw every 5 minutes or when data changes
        new Promise((resolve) => {
          setTimeout(resolve, 1000 * 60 * 5);
          emitter.once("change", resolve);
        }),
      ]);
    }
  }

  await dayview(args);
}
