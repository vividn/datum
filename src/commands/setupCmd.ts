import { Argv } from "yargs";
import { BaseDatumArgs } from "../input/baseYargs";
import { setupDatumViews } from "../views/setupDatumViews";
import { connectDb } from "../auth/connectDb";
import { Show } from "../output/output";

export const command = "setup";
export const desc = "setup the database for use with datum";

export type SetupCmdArgs = BaseDatumArgs & {
  projectDir?: string;
};

export function builder(yargs: Argv): Argv {
  return yargs.options({
    "project-dir": {
      describe:
        "where to look for additional DatumViews to setup in the database",
      type: "string",
    },
  });
}

export async function setupCmd(args: SetupCmdArgs): Promise<void> {
  const db = connectDb(args);
  const show: Show = args.showAll ? Show.All : args.show ?? Show.None;
  await setupDatumViews({ db, show: show, projectDir: args.projectDir });
}
