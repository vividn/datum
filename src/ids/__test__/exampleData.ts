import { DatumData, DatumMetadata } from "../../documentControl/DatumDocument";

export const exampleData: DatumData = {
  foo: "abc",
  bar: "def",
  complex: { data: "nested" },
  array: ["various", 2, "data"],
  num: 3,
  "wei%rd": "da%ta",
};
export const exampleOccurTime = "2020-11-09T00:35:10.000Z";
export const exampleDataOccur = {
  occurTime: exampleOccurTime,
  occurUtcOffset: 1,
  ...exampleData,
};
export const exampleDataOccurField: DatumData = {
  ...exampleDataOccur,
  field: "main",
};
export const exampleMeta: DatumMetadata = {
  createTime: "2020-11-09T00:40:12.544Z",
  modifyTime: "2020-11-09T00:40:12.544Z",
  random: 0.7368733800261729,
  humanId: "mqp4znq4cvp3qnj74fgi9",
};
