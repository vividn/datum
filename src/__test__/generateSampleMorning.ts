import { endCmd } from "../commands/endCmd";
import { occurCmd } from "../commands/occurCmd";
import { startCmd } from "../commands/startCmd";
import { switchCmd } from "../commands/switchCmd";
import { popNow, pushNow, setNow } from "./test-utils";

export async function generateSampleMorning(date: string): Promise<void> {
  pushNow(`8:30 ${date}`);
  await endCmd("sleep");
  setNow("+5");
  await startCmd("sleep -y -t 2320");
  setNow("8:40");
  await switchCmd("project german");
  await switchCmd("text fiction_book");
  setNow("9:10");
  await endCmd("project");
  await endCmd("text");
  setNow("9:30");
  await switchCmd("environment outside");
  await startCmd("stretch");
  setNow("+7");
  await endCmd("stretch");
  await startCmd("run");
  setNow("+30");
  await endCmd("run distance=5.4");
  await startCmd("stretch");
  setNow("+8");
  await endCmd("stretch");
  await switchCmd("environment home");
  setNow("+3");
  await occurCmd("pushup amount=10");
  setNow("10:30");
  await switchCmd("environment outside 5");
  setNow("11");
  await occurCmd("caffeine amount=100 -c coffee");
  popNow();
}
