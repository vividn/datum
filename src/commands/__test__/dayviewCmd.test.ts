import {
  mockedLogLifecycle,
  mockSpecs,
  testDbLifecycle,
} from "../../__test__/test-utils";
import { generateSampleMorning } from "../../__test__/generateSampleMorning";
import { setupCmd } from "../setupCmd";
import { dayviewCmd } from "../dayviewCmd";
import fs from "fs";

describe("dayview", () => {
  const dbName = "dayview_test";
  testDbLifecycle(dbName);
  mockedLogLifecycle();
  mockSpecs();

  beforeEach(async () => {
    setupCmd({});
  });

  it("consistently generates a sample morning svg", async () => {
    await generateSampleMorning("2024-11-20");
    await dayviewCmd({
      endDate: "2024-11-20",
      nDays: 1,
      width: 1000,
      dayHeight: 100,
      outputFile: "/tmp/sample_morning.svg",
    });
    const svg = fs.readFileSync("/tmp/sample_morning.svg", "utf8");
    expect(svg).toMatchSnapshot();
    // write svg to snapshot folder too for github viewing
    fs.writeFileSync(__dirname + "/__snapshots__/sample_morning.svg", svg);
  });
});
