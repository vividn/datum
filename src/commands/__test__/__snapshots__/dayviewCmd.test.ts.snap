// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dayview consistently generates a sample morning svg 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg height="119" width="1000" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg">
    <defs>
        <symbol id="warning-icon">
            <!-- Uploaded to: SVG Repo, www.svgrepo.com, Generator: SVG Repo Mixer Tools -->
            <svg xmlns="http://www.w3.org/2000/svg" fill="#ff0000" viewBox="0 0 24 24" id="warning-alt" data-name="Flat Line" class="icon flat-line">
                <path id="secondary" d="M10.25,4.19,2.63,18a2,2,0,0,0,1.75,3H19.62a2,2,0,0,0,1.75-3L13.75,4.19A2,2,0,0,0,10.25,4.19Z" style="fill: #ffff00; stroke-width: 2;"/>
                <line id="primary-upstroke" x1="12.05" y1="17" x2="11.95" y2="17" style="fill: none; stroke: #ff0000; stroke-linecap: round; stroke-linejoin: round; stroke-width: 2.5;"/>
                <path id="primary" d="M12,9v4M10.25,4.19,2.63,18a2,2,0,0,0,1.75,3H19.62a2,2,0,0,0,1.75-3L13.75,4.19A2,2,0,0,0,10.25,4.19Z" style="fill: none; stroke: #ff0000; stroke-linecap: round; stroke-linejoin: round; stroke-width: 2;"/>
            </svg>
        </symbol>
        <style>
            svg { overflow: visible; }
        </style>
    </defs>
    <rect fill="black" height="100%" width="100%"/>
    <svg height="115" width="996" y="2" x="2" class="plot">
        <g transform="translate(80, 0)">
            <g transform="translate(0, 0)">
                <text font-size="20" text-anchor="end" fill="white" dy="0.35em" y="16.666666666666668" x="-0.5em">
                    Wed
                </text>
                <text font-size="20" text-anchor="end" fill="white" dy="0.35em" y="50" x="-0.5em">
                    Nov 20
                </text>
                <text font-size="20" text-anchor="end" fill="white" dy="0.35em" y="83.33333333333333" x="-0.5em">
                    2024
                </text>
            </g>
        </g>
        <svg x="80" height="100" width="916" class="dataArea">
            <svg y="0" class="day 2024-11-20">
                <svg class="timechunk">
                    <svg height="12" width="916" y="5.209594737738371" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="text">
                        <rect fill="#06de5f" height="100%" width="19.08333355420524" y="0" x="330.7777816062243" class="text fiction_book block"/>
                    </svg>
                    <svg height="12" width="916" y="21.72283353842795" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="project">
                        <rect fill="#a6d414" height="100%" width="19.08333355420524" y="0" x="330.7777816062243" class="project german block"/>
                    </svg>
                    <svg height="12" width="916" y="43.42104974761605" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="stretch">
                        <rect fill="#7e8494" height="100%" width="4.452777829314584" y="0" x="362.5833375298997" class="stretch true block"/>
                        <rect fill="#7e8494" height="100%" width="5.088888947788064" y="0" x="386.11944891341955" class="stretch true block"/>
                    </svg>
                    <svg height="12" width="916" y="51.94052830897271" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="pushup">
                        <circle fill="#9453e7" r="3.1805555923675417" cy="6" cx="393.1166712166281" class="pushup true point"/>
                    </svg>
                    <svg height="12" width="916" y="58.52794650103896" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="run">
                        <rect fill="#a53108" height="100%" width="19.08333355420524" y="0" x="367.0361153592143" class="run true block"/>
                    </svg>
                    <svg height="12" width="916" y="72.89816309325397" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="sleep">
                        <rect fill="#c9fab3" height="100%" width="324.41667042148924" y="0" x="0" class="sleep true block"/>
                        <use field="sleep" class="sleep error" height="9.541666777102625" width="9.541666777102625" y="1.2291666114486874" x="319.64583703293795" href="#warning-icon"/>
                    </svg>
                    <svg height="12" width="916" y="84.29078911989927" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="caffeine">
                        <circle fill="#e724f8" r="3.1805555923675417" cy="6" cx="419.8333381925155" class="caffeine true point"/>
                    </svg>
                    <svg height="12" width="916" y="85.01698424201459" x="0" preserveAspectRatio="none" viewBox="0 0 916 12" class="environment">
                        <rect fill="#c30163" height="100%" width="28.625000331307888" y="0" x="362.5833375298997" class="environment outside block"/>
                        <rect fill="#106a6c" height="100%" width="6.36111118473508" y="0" x="391.2083378612076" class="environment home block"/>
                        <rect fill="#c30163" height="100%" width="3.1805555923675684" y="0" x="397.5694490459427" class="environment outside block"/>
                        <rect fill="#106a6c" height="100%" width="515.2499953616898" y="0" x="400.75000463831026" class="environment home block"/>
                    </svg>
                </svg>
            </svg>
        </svg>
        <g text-anchor="middle" font-family="sans-serif" font-size="10" fill="none" transform="translate(80, 100)">
            <path d="M0.5,6V0.5H916.5V6" stroke="white" class="domain"/>
            <g transform="translate(0.5,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    00
                </text>
            </g>
            <g transform="translate(38.666666666666664,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    01
                </text>
            </g>
            <g transform="translate(76.83333333333333,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    02
                </text>
            </g>
            <g transform="translate(115,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    03
                </text>
            </g>
            <g transform="translate(153.16666666666666,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    04
                </text>
            </g>
            <g transform="translate(191.33333333333334,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    05
                </text>
            </g>
            <g transform="translate(229.5,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    06
                </text>
            </g>
            <g transform="translate(267.6666666666667,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    07
                </text>
            </g>
            <g transform="translate(305.8333333333333,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    08
                </text>
            </g>
            <g transform="translate(344,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    09
                </text>
            </g>
            <g transform="translate(382.1666666666667,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    10
                </text>
            </g>
            <g transform="translate(420.3333333333333,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    11
                </text>
            </g>
            <g transform="translate(458.5,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    12
                </text>
            </g>
            <g transform="translate(496.66666666666663,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    13
                </text>
            </g>
            <g transform="translate(534.8333333333334,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    14
                </text>
            </g>
            <g transform="translate(573,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    15
                </text>
            </g>
            <g transform="translate(611.1666666666666,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    16
                </text>
            </g>
            <g transform="translate(649.3333333333334,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    17
                </text>
            </g>
            <g transform="translate(687.5,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    18
                </text>
            </g>
            <g transform="translate(725.6666666666666,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    19
                </text>
            </g>
            <g transform="translate(763.8333333333334,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    20
                </text>
            </g>
            <g transform="translate(802,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    21
                </text>
            </g>
            <g transform="translate(840.1666666666666,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    22
                </text>
            </g>
            <g transform="translate(878.3333333333334,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    23
                </text>
            </g>
            <g transform="translate(916.5,0)" opacity="1" class="tick">
                <line y2="6" stroke="white"/>
                <text style="font-size:9px" stroke="white" dy="0.71em" y="9" fill="white">
                    00
                </text>
            </g>
        </g>
        <g transform="translate(80, 0)">
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="0" x1="0" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="38.166666666666664" x1="38.166666666666664" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="76.33333333333333" x1="76.33333333333333" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="114.5" x1="114.5" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="152.66666666666666" x1="152.66666666666666" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="190.83333333333334" x1="190.83333333333334" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="229" x1="229" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="267.1666666666667" x1="267.1666666666667" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="305.3333333333333" x1="305.3333333333333" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="343.5" x1="343.5" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="381.6666666666667" x1="381.6666666666667" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="419.8333333333333" x1="419.8333333333333" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="458" x1="458" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="496.16666666666663" x1="496.16666666666663" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="534.3333333333334" x1="534.3333333333334" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="572.5" x1="572.5" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="610.6666666666666" x1="610.6666666666666" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="648.8333333333334" x1="648.8333333333334" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="687" x1="687" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="725.1666666666666" x1="725.1666666666666" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="763.3333333333334" x1="763.3333333333334" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="801.5" x1="801.5" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="839.6666666666666" x1="839.6666666666666" class="gridline"/>
            <line stroke-opacity="0.1" stroke="white" y2="100" y1="0" x2="877.8333333333334" x1="877.8333333333334" class="gridline"/>
            <line stroke-opacity="0.4" stroke="white" y2="100" y1="0" x2="916" x1="916" class="gridline"/>
        </g>
    </svg>
    <use height="20" width="20" y="99" x="0" href="#warning-icon"/>
    <rect fill="red" height="20" width="40" y="99" x="25"/>
    <text text-anchor="start" fill="black" dy="0.35em" y="109" x="25">
        sleep
    </text>
</svg>"
`;
